packages/compiler-core/src/transforms/hoistStatic.ts

```diff
@@ -6,12 +6,24 @@ import {
   ElementCodegenNode,
   PlainElementNode,
   ComponentNode,
-  TemplateNode
+  TemplateNode,
+  ElementNode
 } from '../ast'
 import { TransformContext } from '../transform'
 import { APPLY_DIRECTIVES } from '../runtimeHelpers'
 import { PatchFlags } from '@vue/shared'
-import { isSlotOutlet } from '../utils'
+import { isSlotOutlet, findProp } from '../utils'
+
+function hasDynamicKey(node: ElementNode) {
+  const keyProp = findProp(node, 'key')
+  return (
+    keyProp &&
+    keyProp.type === NodeTypes.DIRECTIVE &&
+    keyProp.exp &&
+    keyProp.exp.type === NodeTypes.SIMPLE_EXPRESSION &&
```

Comment:
This check and the next line are unnecessary, because a directive node is guaranteed to have a dynamic exp.

Reply:
Thank you, that was not obvious for me from type definition, changed

Reply:
Actually I believe we can remove `keyProp.exp` existense check also because this will throw `X_V_BIND_NO_EXPRESSION` compiler error. Is this correct?